src/mlpack/methods/ann/image_functions/bilinear_function.hpp

```diff
@@ -0,0 +1,170 @@
+
+    for (size_t i = 0; i < output.n_rows; i++)
+    {
+      for (size_t j = 0; j < output.n_cols; j++)
+      {
+        rOrigin = std::floor(i * scaleRow);
+        cOrigin = std::floor(j * scaleCol);
+
+        if (rOrigin > input.n_rows - 2)
+          rOrigin = input.n_rows - 2;
+        if (cOrigin > input.n_cols - 2)
+          cOrigin = input.n_cols - 2;
+        
+
+        double deltaR = i * scaleRow - rOrigin;
+        double deltaC = j * scaleCol - cOrigin;
+        coeff1 = (1 - deltaR) * (1 - deltaC);
+        coeff2 = deltaR * (1 - deltaC);
+        coeff3 = (1 - deltaR) * deltaC;
+        coeff4 = deltaR * deltaC;
+
+
+        output(i, j) =  input(cOrigin * input.n_rows + rOrigin) * coeff1 +
+                        input(cOrigin * input.n_rows + rOrigin + 1) * coeff2 +
+                        input((cOrigin + 1) * input.n_rows + rOrigin) * coeff3 +
+                        input((cOrigin + 1) * input.n_rows + rOrigin+1) * coeff4;
+      }
+    }
+  }
+
+  /**
+   * DownSample the given input.
+   *
+   * @param input The input matrix
+   * @param output The resulting down-sampled output image.
+   */
+  template<typename eT>
+  void DownSample(const arma::Mat<eT>& input, arma::Mat<eT>& output)
+  {
+
+    if (output.is_empty() && (output.n_cols != inColSize || output.n_rows != inRowSize))
+      output = arma::resize(output, inRowSize, inColSize);
+
+    if (input.n_rows == output.n_rows && input.n_cols == output.n_cols)
+    {
+      output = input;
+    }
+    else
+    {
+      scaleRow = (double)(input.n_rows - 1) / output.n_rows;
+      scaleCol = (double)(input.n_cols - 1) / output.n_cols;
+
+      for (size_t i = 0; i < output.n_rows; i++)
+        for (size_t j = 0; j < output.n_cols; j++)
+        {
+          rOrigin = std::floor(i * scaleRow);
+          cOrigin = std::floor(j * scaleCol);
+
+          if (rOrigin > input.n_rows - 2)
+            rOrigin = input.n_rows - 2;
+          if (cOrigin > input.n_cols - 2)
+            cOrigin = input.n_cols - 2;
+
+          double deltaR = i * scaleRow - rOrigin;
+          double deltaC = j * scaleCol - cOrigin;
+          coeff1 = (1 - deltaR) * (1 - deltaC);
+          coeff2 = deltaR * (1 - deltaC);
+          coeff3 = (1 - deltaR) * deltaC;
+          coeff4 = deltaR * deltaC;
+
+          output(i, j) =  input(cOrigin * input.n_rows + rOrigin) * coeff1 +
+              input(cOrigin * input.n_rows + rOrigin + 1) * coeff2 +
+              input((cOrigin + 1) * input.n_rows + rOrigin) * coeff3 +
+              input((cOrigin + 1) * input.n_rows + rOrigin+1) * coeff4;
+      }
+    }
+  }
+
+ private:
+  //! Locally stored row size of the input.
+  const size_t inRowSize;
+  //! Locally stored column size of the input.
+  const size_t inColSize;
+  //! Locally stored row size of the output.
+  const size_t outRowSize;
+  //! Locally stored column size of the input.
+  const size_t outColSize;
+
+  //! Locally stored scaling factor along row.
```

Comment:
Well, the variables are used by both the `Upsample` and the `DownSample` function. So rather than creating stack variables, it is better to just create heap variables. 
Could you tell the benefits of using stack variable here?